fix(compiler): adding the inert property to the "SCHEMA" array

physicx237 opened the pull request:
This change allows template binding "inert" attribute with the following syntax: [inert]="isInert"

Fixes #51879

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #51879

## What is the new behavior?
Possibility of template binding of the inert attribute with the following syntax: [inert]="isInert"

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

JeanMeche commented:
Cf the file header

> //                       DO NOT EDIT THIS DOM SCHEMA WITHOUT A SECURITY REVIEW!
> //
> // Newly added properties must be security reviewed and assigned an appropriate SecurityContext in
> // dom_security_schema.ts. Reach out to mprobst & rjamet for details.

I'm marking it as blocked until this is done.

JoostK commented on packages/compiler/src/schema/dom_element_schema_registry.ts:
```diff
@@ -78,7 +78,7 @@ const SCHEMA: string[] = [
       /* added manually to avoid breaking changes */
       ',*message,*mozfullscreenchange,*mozfullscreenerror,*mozpointerlockchange,*mozpointerlockerror,*webglcontextcreationerror,*webglcontextlost,*webglcontextrestored',
-  '[HTMLElement]^[Element]|accessKey,autocapitalize,!autofocus,contentEditable,dir,!draggable,enterKeyHint,!hidden,innerText,inputMode,lang,nonce,*abort,*animationend,*animationiteration,*animationstart,*auxclick,*beforexrselect,*blur,*cancel,*canplay,*canplaythrough,*change,*click,*close,*contextmenu,*copy,*cuechange,*cut,*dblclick,*drag,*dragend,*dragenter,*dragleave,*dragover,*dragstart,*drop, ...
+  '[HTMLElement]^[Element]|accessKey,autocapitalize,!autofocus,contentEditable,dir,!draggable,enterKeyHint,!hidden,!inert,innerText,inputMode,lang,nonce,*abort,*animationend,*animationiteration,*animationstart,*auxclick,*beforexrselect,*blur,*cancel,*canplay,*canplaythrough,*change,*click,*close,*contextmenu,*copy,*cuechange,*cut,*dblclick,*drag,*dragend,*dragenter,*dragleave,*dragover,*dragstart ...
```
This code is generated using

https://github.com/angular/angular/blob/c07805612f77ff59104470a40b86bdd28a93bc61/packages/compiler/test/schema/dom_element_schema_registry_spec.ts#L186-L194

by uncommenting that console.log, changing the `it` to `fit` and running:

```
yarn bazel test packages/compiler/test:test_web_chromium --test_output=streamed
```

or possibly

```
yarn bazel run packages/compiler/test:test_web_debug
```

to run it using your local (Chrome) browser, which may be newer than Chrome 114 as used in the tests.

physicx237 replied:
Yes, inert is the only addition.

JoostK commented on packages/compiler/src/schema/dom_security_schema.ts:
```diff
@@ -31,6 +31,7 @@ export function SECURITY_SCHEMA(): {[k: string]: SecurityContext} {
       'iframe|srcdoc',
       '*|innerHTML',
       '*|outerHTML',
+      '*|inert',
```
I don't think this is needed; `inert` is not a property that needs an HTML security context.

bjarkler replied:
+1, this should be removed.

physicx237 commented:
JeanMeche said:
"See file header

// DO NOT EDIT THIS HOUSE DIAGRAM WITHOUT A SECURITY CHECK!
//
// Newly added properties must be security checked and match the corresponding SecurityContext in
// dom_security_schema.ts. More information on the mprost & rjamet website.

I mark it as blocked until this is done."

Should I leave the "inert" attribute in the security context or is it not needed?

JoostK commented:
The intent of that comment is to make people aware of _potential_ security implications of the schema definitions. For example, Angular must sanitize data when binding to the `innerHTML` property, but not when binding to the `inert` property since that will only be interpreted as boolean value, _not_ rendered as HTML.

Hence, changes to the schema require a security review to avoid mistakenly introducing a vulnerability in the framework. For `inert` though, I cannot think of a reason why it would be security sensitive.

physicx237 commented:
> Cf the file header
> 
> > //                       DO NOT EDIT THIS DOM SCHEMA WITHOUT A SECURITY REVIEW!
> > //
> > // Newly added properties must be security reviewed and assigned an appropriate SecurityContext in
> > // dom_security_schema.ts. Reach out to mprobst & rjamet for details.
> 
> I'm marking it as blocked until this is done. 

How to reach out to rjamet & mbprost to details?

jelbourn commented:
Ah, that comment is out of date (mbprost & rjamet aren't involved with Angular any longer). @bjarkler I believe `inert` should be completely safe here as a boolean attribute, but I'll add you for FYI anyway since it's part of the process for changing the schema. 

LGTM insofar as adding `inert` in general, just need to address the comments from @JoostK

bjarkler commented:
LGTM, thanks!

physicx237 commented:
Thanks for your feedback!

JoostK commented:
Could you please squash the commits? The revert won't be auto squashed otherwise

physicx237 commented:
Yes, sure

JoostK commented:
Looks like something went wrong, there's now five commits.

physicx237 committed:
fix(compiler): adding the inert property to the "SCHEMA" array

This change allows template binding "inert" attribute with the following syntax: [inert]="isInert"

Fixes #51879

fixup! fix(compiler): adding the inert property to the "SCHEMA" array

revert: "fixup! fix(compiler): adding the inert property to the "SCHEMA" array"

This reverts commit b637b7ce646e8bab2f585339028a84018e8ea982.

This commit is being reverted because the inert property is safe as a boolean attribute

JoostK reviewed (approved):
Thanks for the updates!

physicx237 commented:
And thank you!

jelbourn reviewed (approved):
LGTM

Reviewed-for: fw-security

physicx237 commented:
Hello @alxhub, @pkozlowski-opensource and @josephperrott! Can you review my code?

physicx237 commented:
Hello @alxhub, @pkozlowski-opensource and @josephperrott! Can you review my code?

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: fw-security

alxhub commented:
This PR was merged into the repository by commit dba3e0b5aa73d49a6caad4ad474e0212486fe98a.
